lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -4,155 +4,208 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
-
 #include "hermes/Platform/Intl/PlatformIntl.h"
 
 #import <Foundation/Foundation.h>
 
+NSString *u16StringToNSString(std::u16string src) {
+  auto size = src.size();
+  if (size == 0) {
+    return nil;
+  }
+  auto cString = (const unichar*)src.c_str();
+  if (cString == nullptr) {
+    return nil;
+  }
+  return [NSString stringWithCharacters:cString length:size];
+}
+
+std::u16string nsStringToU16String(NSString *src) {
+  auto result = std::u16string();
+  auto constexpr divisor = sizeof(char16_t) / sizeof(char);
+  if ([src canBeConvertedToEncoding:NSUTF16StringEncoding]) {
+    auto size = [src lengthOfBytesUsingEncoding:NSUTF16StringEncoding] + 1; // for NULL terminator;
+    auto buffer = (char *)malloc(size);
```

Comment:
Instead of setting up a temporary buffer, we can resize the string upfront and use that as the buffer directly.